core/drivers/crypto/caam/utils/utils_dmaobj.c

```diff
@@ -0,0 +1,1395 @@
+	if (!priv->nb_sgtbuf) {
+		DMAOBJ_TRACE("From SGT/Buffer not prepared");
+		ret = TEE_ERROR_NO_DATA;
+		goto end;
+	}
+
+	retstatus =
+		caam_sgt_derive(&obj->sgtbuf, &from->sgtbuf, offset, length);
+
+	ret = caam_status_to_tee_result(retstatus);
+
+end:
+	DMAOBJ_TRACE("Object returns 0x%" PRIx32, ret);
+	return ret;
+}
+
+TEE_Result caam_dmaobj_prepare(struct caamdmaobj *input,
+			       struct caamdmaobj *output, size_t min_size)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	struct priv_dmaobj *priv_input = NULL;
+	struct priv_dmaobj *priv_output = NULL;
+	size_t max_alloc_input = 0;
+	size_t max_alloc_output = 0;
+	uint32_t exceptions = 0;
+	bool try_input_alloc = false;
+	bool try_output_alloc = false;
+
+	if (!input && !output) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto end;
+	}
+
+	if ((input && !input->priv) || (output && !output->priv)) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto end;
+	}
+
+	DMAOBJ_TRACE("input=%p - output=%p - min=%zu", input, output, min_size);
+
+	if (input) {
+		priv_input = input->priv;
+		DMAOBJ_TRACE("Input DMA buffer size require %zu",
+			     priv_input->dmabuf.require);
+		max_alloc_input =
+			MIN(priv_input->dmabuf.require, MAX_BUFFER_ALLOC_SIZE);
+		if (max_alloc_input > 1024)
+			max_alloc_input = ROUNDDOWN(max_alloc_input, 1024);
+
+		if (max_alloc_input)
+			try_input_alloc = true;
+	}
+
+	if (output) {
+		priv_output = output->priv;
+		DMAOBJ_TRACE("Output DMA buffer size require %zu",
+			     priv_output->dmabuf.require);
+		max_alloc_output =
+			MIN(priv_output->dmabuf.require, MAX_BUFFER_ALLOC_SIZE);
+		if (max_alloc_output > 1024)
+			max_alloc_output = ROUNDDOWN(max_alloc_output, 1024);
+
+		if (max_alloc_output)
+			try_output_alloc = true;
+	}
+
+	/*
+	 * If require:
+	 * - Try to allocate the input big buffer.
+	 * - Then Try to allocate the output big buffer.
+	 *
+	 * Loop while by dividing allocation size by 2 with a minimum of
+	 * @min_size.
+	 * If allocation with minimum size for both is failing,
+	 * returns in error.
+	 */
+	exceptions = cpu_spin_lock_xsave(&memlock);
+
+retry:
+	DMAOBJ_TRACE("Allocation input %zu output %zu", max_alloc_input,
+		     max_alloc_output);
+	if (try_input_alloc) {
+		ret = try_allocate_dmabuf(priv_input, max_alloc_input);
+		if (!ret)
+			try_input_alloc = false;
+	} else {
+		ret = TEE_SUCCESS;
+	}
+
+	if (try_output_alloc && !ret) {
+		ret = try_allocate_dmabuf(priv_output, max_alloc_output);
+		if (!ret)
+			try_output_alloc = false;
+	}
+
+	if (ret) {
+		if (max_alloc_input) {
+			if (max_alloc_input > min_size) {
+				max_alloc_input =
+					MAX(min_size, max_alloc_input / 2);
+				try_input_alloc = true;
+			} else {
+				try_input_alloc = false;
+			}
+		}
+
+		if (max_alloc_output) {
+			if (max_alloc_output > min_size) {
+				max_alloc_output =
+					MAX(min_size, max_alloc_output / 2);
+				try_output_alloc = true;
+			} else {
+				try_output_alloc = false;
+			}
+		}
+
+		if (try_input_alloc || try_output_alloc)
+			goto retry;
```

Comment:
Please avoid creating loops with `goto`.

Reply:
I will replace it with a ```while (try_input_alloc || try_output_alloc) {}``` loop